File: apps/demo-be-kotlin-ktor/src/test/kotlin/com/organiclever/demoktkt/unit/steps/UnitAdminSteps.kt
package com.organiclever.demoktkt.unit.steps

import io.cucumber.java.en.Then
import io.cucumber.java.en.When
import org.junit.jupiter.api.Assertions.assertTrue

class UnitAdminSteps {

  @When("the admin sends GET \\/api\\/v1\\/admin\\/users")
  fun theAdminSendsGetAdminUsers() {
    val token = UnitTestWorld.accessTokens["superadmin"] ?: error("admin token not stored")
    val (status, body) = UnitHttpHelper.get("/api/v1/admin/users", token)
    UnitTestWorld.lastResponseStatus = status
    UnitTestWorld.lastResponseBody = body
  }

  @When("^the admin sends GET /api/v1/admin/users\\?email=([^@]+@[^\\s]+)$")
  fun theAdminSendsGetAdminUsersWithEmailFilter(email: String) {
    val token = UnitTestWorld.accessTokens["superadmin"] ?: error("admin token not stored")
    val (status, body) = UnitHttpHelper.get("/api/v1/admin/users?email=$email", token)
    UnitTestWorld.lastResponseStatus = status
    UnitTestWorld.lastResponseBody = body
  }

  @When(
    "^the admin sends POST /api/v1/admin/users/\\{alice_id\\}/disable with body \\{ \"reason\": \"([^\"]+)\" \\}$"
  )
  fun theAdminSendsPostDisableAliceWithBody(reason: String) {
    val aliceId = UnitTestWorld.userIds["alice"] ?: error("alice id not stored")
    val adminToken = UnitTestWorld.accessTokens["superadmin"] ?: error("admin token not stored")
    val body = """{"reason":"$reason"}"""
    val (status, respBody) =
      UnitHttpHelper.post("/api/v1/admin/users/$aliceId/disable", body, adminToken)
    UnitTestWorld.lastResponseStatus = status
    UnitTestWorld.lastResponseBody = respBody
  }

  @When("^the admin sends POST /api/v1/admin/users/\\{alice_id\\}/enable$")
  fun theAdminSendsPostEnableAlice() {
    val aliceId = UnitTestWorld.userIds["alice"] ?: error("alice id not stored")
    val adminToken = UnitTestWorld.accessTokens["superadmin"] ?: error("admin token not stored")
    val (status, body) = UnitHttpHelper.post("/api/v1/admin/users/$aliceId/enable", "", adminToken)
    UnitTestWorld.lastResponseStatus = status
    UnitTestWorld.lastResponseBody = body
  }

  @When("^the admin sends POST /api/v1/admin/users/\\{alice_id\\}/force-password-reset$")
  fun theAdminSendsPostForcePasswordResetAlice() {
    val aliceId = UnitTestWorld.userIds["alice"] ?: error("alice id not stored")
    val adminToken = UnitTestWorld.accessTokens["superadmin"] ?: error("admin token not stored")
    val (status, body) =
      UnitHttpHelper.post("/api/v1/admin/users/$aliceId/force-password-reset", "", adminToken)
    UnitTestWorld.lastResponseStatus = status
    UnitTestWorld.lastResponseBody = body
  }

  @Then("^the response body should contain at least one user with \"email\" equal to \"([^\"]+)\"$")
  fun theResponseBodyShouldContainAtLeastOneUserWithEmail(email: String) {
    val body = UnitTestWorld.lastResponseBody
    assertTrue(body.contains(email), "Expected email '$email' in response: $body")
  }
}

File: apps/a-demo-be-csharp-aspnetcore/tests/ADemoBeCsas.Tests/Unit/Steps/UnitAdminSteps.cs
using System.Text.Json;
using FluentAssertions;
using Reqnroll;
using Xunit;

namespace ADemoBeCsas.Tests.Unit.Steps;

/// <summary>
/// Unit-level step bindings for admin user management scenarios.
/// Mirrors <c>Integration.Steps.AdminSteps</c> but scoped to scenarios tagged
/// <c>@unit</c> and uses <see cref="UnitServiceLayer"/> with in-memory repositories.
/// </summary>
[Binding]
[Trait("Category", "Unit")]
[Scope(Tag = "unit")]
public class UnitAdminSteps(UnitServiceLayer svc, UnitSharedState state, UnitAuthSteps auth)
{
    // ─────────────────────────────────────────────────────────────
    // When steps
    // ─────────────────────────────────────────────────────────────

    [When(@"^the admin sends GET /api/v1/admin/users$")]
    public async Task WhenAdminListsUsers()
    {
        state.LastResponse = await svc.AdminListUsersAsync(auth._adminToken, page: 1, size: 20);
    }

    [When(@"^the admin sends GET /api/v1/admin/users\?search=alice@example\.com$")]
    public async Task WhenAdminSearchesByEmail()
    {
        state.LastResponse = await svc.AdminListUsersAsync(
            auth._adminToken,
            page: 1,
            size: 20,
            search: "alice@example.com"
        );
    }

    [When(@"^the admin sends POST /api/v1/admin/users/\{alice_id\}/disable with body \{ ""reason"": ""([^""]+)"" \}$")]
    public async Task WhenAdminDisablesAlice(string reason)
    {
        var aliceId = state.LastCreatedId ?? auth._aliceId;
        aliceId.Should().NotBeNull("alice's ID should be known");
        state.LastResponse = await svc.AdminDisableUserAsync(auth._adminToken, aliceId!.Value);
    }

    [When(@"^the admin sends POST /api/v1/admin/users/\{alice_id\}/enable$")]
    public async Task WhenAdminEnablesAlice()
    {
        var aliceId = state.LastCreatedId ?? auth._aliceId;
        aliceId.Should().NotBeNull("alice's ID should be known");
        state.LastResponse = await svc.AdminEnableUserAsync(auth._adminToken, aliceId!.Value);
    }

    [When(@"^the admin sends POST /api/v1/admin/users/\{alice_id\}/unlock$")]
    public async Task WhenAdminUnlocksAlice()
    {
        var aliceId = state.LastCreatedId ?? auth._aliceId;
        aliceId.Should().NotBeNull("alice's ID should be known");
        state.LastResponse = await svc.AdminUnlockUserAsync(auth._adminToken, aliceId!.Value);
    }

    [When(@"^the admin sends POST /api/v1/admin/users/\{alice_id\}/force-password-reset$")]
    public async Task WhenAdminForcePasswordReset()
    {
        var aliceId = state.LastCreatedId ?? auth._aliceId;
        aliceId.Should().NotBeNull("alice's ID should be known");
        state.LastResponse = await svc.AdminForcePasswordResetAsync(
            auth._adminToken,
            aliceId!.Value
        );
        if (state.LastResponse.IsSuccess)
        {
            using var doc = JsonDocument.Parse(state.LastResponse.Body);
            if (doc.RootElement.TryGetProperty("token", out var rt))
            {
                state.LastResetToken = rt.GetString();
            }
        }
    }

    // ─────────────────────────────────────────────────────────────
    // Then steps
    // ─────────────────────────────────────────────────────────────

    [Then(@"^the response body should contain at least one user with ""email"" equal to ""([^""]+)""$")]
    public void ThenResponseContainsUserWithEmail(string email)
    {
        state.LastResponse.Should().NotBeNull();
        var body = state.LastResponse!.Body;
        body.Should().Contain(email, $"Expected user with email '{email}' in: {body}");
    }
}
